Implement `document.scrollingElement`

JimmyDdotEXE opened the pull request:
<!-- Please describe your changes on the following line: -->
This implements `document.scrollingElement` (https://drafts.csswg.org/cssom-view/#dom-document-scrollingelement).

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #35700
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

JimmyDdotEXE committed:
Implement document.scrollingElement

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

JimmyDdotEXE committed:
Update web platform tests for document.scrollingElement

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

simonwuelker reviewed (changes requested):
This looks like a good start!

You seem to be missing part of the spec though:
> For this purpose, a value of overflow:clip on the [the body element](https://html.spec.whatwg.org/multipage/dom.html#the-body-element-2)’s parent element must be treated as overflow:hidden.

Is there a reason that we don't need to care about this part?

simonwuelker commented on components/script_bindings/codegen/Bindings.conf:
```diff
@@ -134,7 +134,7 @@ DOMInterfaces = {
 'Document': {
     'additionalTraits': ["crate::dom::document::DocumentHelpers<Self>"],
-    'canGc': ['Close', 'CreateElement', 'CreateElementNS', 'ImportNode', 'SetTitle', 'Write', 'Writeln', 'CreateEvent', 'CreateRange', 'Open', 'Open_', 'CreateComment', 'CreateAttribute', 'CreateAttributeNS', 'CreateDocumentFragment', 'CreateTextNode', 'CreateCDATASection', 'CreateProcessingInstruction', 'Prepend', 'Append', 'ReplaceChildren', 'SetBgColor', 'SetFgColor', 'Fonts', 'ElementFromPoin ...
+    'canGc': ['Close', 'CreateElement', 'CreateElementNS', 'ImportNode', 'SetTitle', 'Write', 'Writeln', 'CreateEvent', 'CreateRange', 'Open', 'Open_', 'CreateComment', 'CreateAttribute', 'CreateAttributeNS', 'CreateDocumentFragment', 'CreateTextNode', 'CreateCDATASection', 'CreateProcessingInstruction', 'Prepend', 'Append', 'ReplaceChildren', 'SetBgColor', 'SetFgColor', 'Fonts', 'ElementFromPoin ...
```
This should be `GetScrollingElement`. That will give you a `canGc` argument that you can pass to calls in the function.

JimmyDdotEXE replied:
This should be addressed now. Thanks for the feedback!

simonwuelker commented on tests/wpt/meta/css/cssom-view/scroll-overflow-clip-quirks-001.html.ini:
```diff
@@ -1,4 +1,2 @@
 [scroll-overflow-clip-quirks-001.html]
-  [CSSOM scrollingElement reflects the propagated scroll to viewport correctly]
-    expected: FAIL
```
You can remove the test file once it contains no more failing tests

JimmyDdotEXE replied:
Good to know. This is done, and I appreciate the explanation.

Loirooriol commented on components/script/dom/document.rs:
```diff
@@ -5689,6 +5689,32 @@ impl DocumentMethods<crate::DomTypeHolder> for Document {
+                }
+            }
+
+            //Step 1.2. Return null and abort these steps.
```
Nit:
```suggestion
            // Step 1.2. Return null and abort these steps.
```

JimmyDdotEXE committed:
Update document.scrollingElement implementation based on review

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

JimmyDdotEXE committed:
Remove unused wpt configurations after implementing document.scrollingElement

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

JimmyDdotEXE commented on components/script/dom/document.rs:
```diff
@@ -5689,6 +5689,32 @@ impl DocumentMethods<crate::DomTypeHolder> for Document {
+
+        // Step 2. If there is a root element, return the root element and abort these steps.
+        // Step 3. Return null.
+        self.GetFirstElementChild()
```
That has been addressed. Thank you for the advice!

JimmyDdotEXE commented on components/script_bindings/webidls/Document.webidl:
```diff
@@ -204,6 +204,11 @@ partial interface Document {
 
+// https://drafts.csswg.org/cssom-view/#extensions-to-the-document-interface
+partial interface Document {
+  readonly attribute Element? scrollingElement;
```
I added `CaretPosition? caretPositionFromPoint(...);` and commented it out for now.

Both `Element? elementFromPoint (...);` and `sequence<Element> elementsFromPoint (...);` are handled in `DocumentOrShadowRoot.webidl`. Do they need to be included and commented out here as well?

JimmyDdotEXE commented on components/script/dom/document.rs:
```diff
@@ -5689,6 +5689,32 @@ impl DocumentMethods<crate::DomTypeHolder> for Document {
         )
     }
 
+    // https://drafts.csswg.org/cssom-view/#dom-document-scrollingelement
```
This is addressed now.

Is this a standard way to include the definition being implemented? Also, if yes, should there be an issue to update similar comments in `document.rs`?

JimmyDdotEXE replied:
Thank you for the explanation! I'll make sure the follow that for future pull requests.

JimmyDdotEXE commented:
> This looks like a good start!
> 
> You seem to be missing part of the spec though:
> 
> > For this purpose, a value of overflow:clip on the [the body element](https://html.spec.whatwg.org/multipage/dom.html#the-body-element-2)’s parent element must be treated as overflow:hidden.
> 
> Is there a reason that we don't need to care about this part?

I think this section of `Element.is_potentially_scrollable_body(...)` covers that stipulation.
https://github.com/servo/servo/blob/94d12de7fd2f6dcd226566280557a1de480e8adc/components/script/dom/element.rs#L458-L468
I could very well be wrong about that, so please let me know if any additional implementation is necessary.

Also, if that part of the specification needs added as a comment for completeness, please let me know.

simonwuelker commented:
> I think this section of Element.is_potentially_scrollable_body(...) covers that stipulation.

The spec says

> For this purpose, a value of overflow:clip on the [the body element](https://html.spec.whatwg.org/multipage/dom.html#the-body-element-2)’s parent element must be treated as overflow:hidden.

`is_potentially_scrollable_body` calls [`is_scrollable`](https://docs.rs/stylo/latest/src/stylo/values/specified/box.rs.html#1810-1812), which returns true for `overflow: hidden / auto / scroll`, but not for `clip`. So you probably need to add a parameter to `is_potentially_scrollable_body` that indicates whether `overflow: clip` should be considered scrollable or not.

JimmyDdotEXE committed:
Implement parent overflow:clip as overflow:hidden for document.scrollingElement

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

JimmyDdotEXE commented:
> Yeah, please include the full spec text - otherwise people (like me) will inevitably wonder why some parts are missing (:

Alright. That portion of the specification has been added to the comments for that method. Please let me know if it needs clarified in any way.

> `is_potentially_scrollable_body` calls [`is_scrollable`](https://docs.rs/stylo/latest/src/stylo/values/specified/box.rs.html#1810-1812), which returns true for `overflow: hidden / auto / scroll`, but not for `clip`. So you probably need to add a parameter to `is_potentially_scrollable_body` that indicates whether `overflow: clip` should be considered scrollable for the parent element or not.

I've gone ahead and tried to implement this in `Element.is_potentially_scrollable_body(...)`:
https://github.com/servo/servo/blob/c995fc11ec28571dd8d0c44bbbc81e304ae8aab1/components/script/dom/element.rs#L467-L483

I will note that the `css/cssom` and `css/cssom-view` web platform test results don't seem to change with or without this new implementation. I'm not sure if that means I've implemented it incorrectly or not.

I don't feel the most confident about this change, so I would be happy to hear any feedback or advice.

simonwuelker commented on components/script/dom/element.rs:
```diff
@@ -441,7 +442,11 @@ impl Element {
-    pub(crate) fn is_potentially_scrollable_body(&self, can_gc: CanGc) -> bool {
+    pub(crate) fn is_potentially_scrollable_body(
+        &self,
+        clip_as_hidden: bool,
```
nit: I would prefer something like `treat_overflow_clip_as_hidden_on_parent`, since this parameter doesn't change how the overflow property is treated on the element itself.

JimmyDdotEXE replied:
I totally agree. Thank you for catching this!

simonwuelker commented on components/script/dom/element.rs:
```diff
@@ -459,9 +464,21 @@ impl Element {
+                let mut overflow_x = style.get_box().clone_overflow_x();
+                let mut overflow_y = style.get_box().clone_overflow_y();
+
+                // This fulfills 'treat parent element overflow:clip as overflow:hidden' stipulation
+                // from document.scrollingElement specification.
```
nit: *the* stipulation from *the* specification

JimmyDdotEXE replied:
I believe I've addressed this, but please let me know if I missed your intent.

simonwuelker replied:
Thanks (:

simonwuelker reviewed (commented):
> I don't feel the most confident about this change, so I would be happy to hear any feedback or advice.

It looks correct to me. It's possible that existing tests don't cover this specific case, in which case we should add some.
Have you confirmed that the behaviour for overflow:clip on the parent matches that of mainstream engines?

JimmyDdotEXE committed:
document.scrollingElement: comment and parameter adjustment based on review

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

JimmyDdotEXE commented:
> > I don't feel the most confident about this change, so I would be happy to hear any feedback or advice.
> 
> It looks correct to me. It's possible that existing tests don't cover this specific case, in which case we should add some. Have you confirmed that the behaviour for overflow:clip on the parent matches that of mainstream engines?

I've only peaked at their implementations for `scrollingElement` and their respective versions `isPotentiallyScrollable()` so far. I know that's not quite the same as confirming behavior, but it seemed like an easy place to start sanity checking.

I might be missing something sneaky about their implementations, but I don't see where [WebKit](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/dom/Document.cpp#L2316), [Gecko](https://hg.mozilla.org/mozilla-central/file/tip/dom/base/Document.cpp#l14306), or [Blink](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/blink/renderer/core/layout/layout_object.h#1535) are directly handling that portion of the specification. The fact that their respective versions of `isPotentiallyScrollable()` don't accept a parameter for this condition furthers that thought for me.

If confirming behavior seems like the next logical step, how might I go about that?

Also, I realize the links I included for reference are fragile, and that worries me from a historical standpoint. Is there a better way to accomplish a similar goal? If not, is it worth removing them after resolving this conversation to avoid link rot in the future?

simonwuelker commented:
> I might be missing something sneaky about their implementations, but I don't see where [WebKit](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/dom/Document.cpp#L2316), [Gecko](https://hg.mozilla.org/mozilla-central/file/tip/dom/base/Document.cpp#l14306), or [Blink](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/blink/renderer/core/layout/layout_object.h#1535) are directly handling that portion of the specification. The fact that their respective versions of isPotentiallyScrollable() don't accept a parameter for this condition furthers that thought for me.

[Ladybird](https://github.com/LadybirdBrowser/ladybird/blob/6f69a445bd3d2fd5bbd61b3b3cea47df18fa6242/Libraries/LibWeb/DOM/Document.cpp#L5543-L5546) also doesn't seem to bother with it. How odd. 

> If confirming behavior seems like the next logical step, how might I go about that?

I'd construct a test case with `overflow: clip` on the html element and see what other browsers return from `document.scrollingElement`.

> Also, I realize the links I included for reference are fragile, and that worries me from a historical standpoint. Is there a better way to accomplish a similar goal? If not, is it worth removing them after resolving this conversation to avoid link rot in the future?

Do you mean `https://drafts.csswg.org/cssom-view/#extensions-to-the-document-interface`? Sure, if there was a non-draft version of the spec then linking to that would be preferred, but it's better to have potentially outdated links than no links at all.

simonwuelker commented:
https://github.com/w3c/csswg-drafts/issues/5326

simonwuelker commented:
| Servo (with your changes) | Firefox | Chrome | 
| --- | --- | --- |
| ![image](https://github.com/user-attachments/assets/cfc09796-0c8d-45b7-9ed8-2e25e991a0ed) | ![image](https://github.com/user-attachments/assets/74d48119-3ae0-4c0b-882d-7feaba753555) |   ![image](https://github.com/user-attachments/assets/50395173-cbfd-4a56-a9a5-4b6c70ffe52a) |

(a green square means that `document.scrollingElement` is not null)

<details><summary>Test HTML</summary>

```html
<html>

<head>
  <style>
    table {
      border: 1px solid black;
    }
    th {
      width: 50px;
      height: 50px;
    }

    td {
      width: 30px;
      height: 30px;
      margin: 5px;
      border: 1px solid black;
    }

    td.null {
      background-color: red;
    }

    td.nonnull {
      background-color: green;
    }
  </style>
</head>

<body>
  <table id="results"></table>

  <script defer>
    let OVERFLOW_VALUES = ["auto", "hidden", "scroll", "visible", "clip"];

    let header = document.createElement("tr");
    header.innerHTML += "<th></th>";
    OVERFLOW_VALUES.forEach((value) => {
      header.innerHTML += `<th><code><code>&lt;body&gt; ${value}</code></th>`;
    })
    results.appendChild(header);

    for (const html_overflow of OVERFLOW_VALUES) {
      document.documentElement.style.overflow = html_overflow;

      let tr = document.createElement("tr");
      tr.innerHTML = `<th><code>&lt;html&gt; ${html_overflow}</code></th>`;
      for (const body_overflow of OVERFLOW_VALUES) {
        document.body.style.overflow = body_overflow;

        let class_value = document.scrollingElement ? "nonnull" : "null";
        tr.innerHTML += `<td class=\"${class_value}\"></td>`;
      }

      results.appendChild(tr);
    }
  </script>
</body>

</html>
```
</details>

Loirooriol commented:
Red symbolizes a test failure, so please use another color for null since it's sometimes expected.

JimmyDdotEXE commented:
> Do you mean `https://drafts.csswg.org/cssom-view/#extensions-to-the-document-interface`? Sure, if there was a non-draft version of the spec then linking to that would be preferred, but it's better to have potentially outdated links than no links at all.

I actually meant the links I'd included to WebKit, Gecko, and Blink code. I was worried those links wouldn't point to the correct lines of code as those projects and respective files changed over time. I've updated that comment to use permalinks instead, so I don't think that will be an issue now.

JimmyDdotEXE commented:
> It seems that firefox is wrong and chrome (+ servo 🎉 ) are correct.

Wow! This is my first time working on a browser engine, so it's super cool to see that side by side comparison. Thank you for putting that together. Also, thank you for all your help in refining this pull request!

> Do you want to add a web platform test for this behaviour?

Would adding a web platform test be an addition to this repository, or is there a central definition for those that Servo pulls from?

simonwuelker commented:
> Would adding a web platform test be an addition to this repository, or is there a central definition for those that Servo pulls from?

Both, kinda (:
Web platform tests live in https://github.com/web-platform-tests/wpt, and you *can* make a direct pull request there.
But servo also vendors its own copy of them in [`tests/wpt/tests`](https://github.com/servo/servo/tree/main/tests/wpt/tests). These are kept in sync via regular updates, like in https://github.com/servo/servo/pull/36019.

The usual workflow for adding a new test is to add it in the servo repository, then a bot will open an upstream pull request that is automatically merged when the servo PR is merged [^1].
Documentation on how to add a test can be found in https://book.servo.org/hacking/testing.html.

[^1]: https://github.com/web-platform-tests/wpt/pulls?q=is%3Apr+is%3Aopen+label%3Aservo-export

JimmyDdotEXE commented:
> The usual workflow for adding a new test is to add it in the servo repository, then a bot will open an upstream pull request that is automatically merged when the servo PR is merged [1](#user-content-fn-1-54eef2d6910eeee8bff3e7ca0accee2c). Documentation on how to add a test can be found in https://book.servo.org/hacking/testing.html.

Alright, I'm going to try my hand at this. I noticed that the file [`tests/wpt/tests/css/cssom-view/scrollingElement.html`](https://github.com/servo/servo/blob/main/tests/wpt/tests/css/cssom-view/scrollingElement.html) pretty well covers `scrollingElement`. Would you recommend just adding new test(s) to that file or a new test file in this case? Should I add the same test in quicks and no-quircks mode?

simonwuelker commented:
The issue we found above only affects quirks mode, so I'd only test in quirks mode.

I'm not sure what the convention for adding a new test vs modifying an existing test is, but I'd lean towards creating  new one and calling it something like `scrollingElement-quirks.html` and then moving the existing quirks test from `scrollingElement.html` over.

JimmyDdotEXE committed:
document.scrollingElement: update web platform tests for overflow:clip

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#51826) with upstreamable changes.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51826).

JimmyDdotEXE commented:
> The issue we found above only affects quirks mode, so I'd only test in quirks mode.
> 
> I'm not sure what the convention for adding a new test vs modifying an existing test is, but I'd lean towards creating new one and calling it something like `scrollingElement-quirks.html` and then moving the existing quirks test from `scrollingElement.html` over.

Alright. I've just added these lines to [`scrollingElement.html`](https://github.com/servo/servo/blob/main/tests/wpt/tests/css/cssom-view/scrollingElement.html).
https://github.com/servo/servo/blob/134f18d1f74d7a88b0d6f5c2a2d8f7d1f85a4ee2/tests/wpt/tests/css/cssom-view/scrollingElement.html#L38-L48

I think this should cover the bottom row of boxes from the test you put together. Please let me know if you have any feedback.

For now, I chose not to split that test file into two. My thought was to limit the number of changes in this pull request, but I can go ahead and split if you feels that's needed.

JimmyDdotEXE commented:
I rebased this branch to address some merge conflicts. For future reference, does this project have a preferred method of addressing merge conflicts?

Loirooriol commented:
Rebasing seems better

JimmyDdotEXE commented:
> Rebasing seems better

Alright, I'll continue rebasing. Thank you for the guidance.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51826).

JimmyDdotEXE committed:
document.scrollingElement: remove .and_then(DomRoot::downcast::<HTMLBodyElement>)

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51826).

JimmyDdotEXE commented on components/script/dom/document.rs:
```diff
@@ -5907,6 +5907,33 @@ impl DocumentMethods<crate::DomTypeHolder> for Document {
+            // Step 1.1. If the body element exists,
+            if let Some(ref body) = self
+                .GetBody()
+                .and_then(DomRoot::downcast::<HTMLBodyElement>)
```
I think I'd looked around the file to see how other methods got the body element, and that was just the first thing I tried that had worked. It seems to build and test fine without it, so I've removed `.and_then(DomRoot::downcast::<HTMLBodyElement>)` now.

Thank you for the feedback.

Loirooriol commented on components/script/dom/element.rs:
```diff
@@ -2752,7 +2769,7 @@ impl ElementMethods<crate::DomTypeHolder> for Element {
         if doc.GetBody().as_deref() == self.downcast::<HTMLElement>() &&
             doc.quirks_mode() == QuirksMode::Quirks &&
-            !self.is_potentially_scrollable_body(can_gc)
+            !self.is_potentially_scrollable_body(false, can_gc)
```
Passing a plain bool argument with no comment or anything to indicate what it means isn't a good practice.

But `self.is_potentially_scrollable_body(false /* treat_overflow_clip_on_parent_as_hidden */, can_gc)` is kind of annoyingly long, and it's not an input flag in the spec algorithm.

Maybe expose `is_potentially_scrollable_body(can_gc)` and `is_potentially_scrollable_body_for_scrolling_element(can_gc)` and then let them call a private function with the common logic.

simonwuelker replied:
Something I like to do is add an enum like
```rust
enum TreatOverflowClipOnParentAsHidden {
    Yes,
    No,
}
```

Though admittedly that name does feel clunky.

Loirooriol replied:
Yes in C++ I would use the enum and make it default to No if omitted, but rust doesn't have optional functional parameters so it's still tedious to specify that on all calls when it's not really an input of the spec algorithm

JimmyDdotEXE replied:
> Maybe expose `is_potentially_scrollable_body(can_gc)` and `is_potentially_scrollable_body_for_scrolling_element(can_gc)` and then let them call a private function with the common logic.

Thank you for the feedback. This should be addressed now.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51826).

JimmyDdotEXE committed:
document.scrollingElement: split element.is_potentially_scrollable_body()

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51826).

JimmyDdotEXE committed:
document.scrollingElement: update intersection-observer test configs

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51826).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51826).

JimmyDdotEXE committed:
document.scrollingElement: update expected web platform test results

Signed-off-by: JimmyDdotEXE <50691404+JimmyDdotEXE@users.noreply.github.com>

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51826).

simonwuelker commented:
Thanks for updating the remaining tests too. Let's hope we can get this merged now...

JimmyDdotEXE commented:
Oh sweet! This got merged. Thank you @simonwuelker and @Loirooriol for help and guidance getting this ready to merge!
